File: android/app/src/main/java/com/brainbites/BrainBitesTimerModule.java
// android/app/src/main/java/com/brainbites/timer/BrainBitesTimerModule.java
package com.brainbites.timer;

import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;
import java.util.Map;

public class BrainBitesTimerModule extends ReactContextBaseJavaModule implements TimerServiceCallback {
    
    private static final String MODULE_NAME = "BrainBitesTimer";
    private static final String TAG = "BrainBitesTimer";
    private static final String PREFS_NAME = "BrainBitesTimerPrefs";
    
    // Constants for events
    private static final String EVENT_TIMER_UPDATE = "timerUpdate";
    private static final String EVENT_SCREEN_STATE_CHANGED = "screenStateChanged";
    private static final String EVENT_APP_STATE_CHANGED = "appStateChanged";
    
    // Timer state constants
    private static final String KEY_REMAINING_TIME = "remainingTime";
    private static final String KEY_IS_TRACKING = "isTracking";
    private static final String KEY_DEBT_TIME = "debtTime";
    private static final String KEY_LAST_UPDATE = "lastUpdate";
    private static final String KEY_SCREEN_STATE = "screenState";
    private static final String KEY_APP_STATE = "appState";
    
    private final ReactApplicationContext reactContext;
    private ScreenStateReceiver screenStateReceiver;
    private SharedPreferences prefs;
    private boolean isServiceRunning = false;
    private boolean isListening = false;
    
    public BrainBitesTimerModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.prefs = reactContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        initializeScreenStateReceiver();
    }
    
    @NonNull
    @Override
    public String getName() {
        return MODULE_NAME;
    }
    
    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("EVENT_TIMER_UPDATE", EVENT_TIMER_UPDATE);
        constants.put("EVENT_SCREEN_STATE_CHANGED", EVENT_SCREEN_STATE_CHANGED);
        constants.put("EVENT_APP_STATE_CHANGED", EVENT_APP_STATE_CHANGED);
        return constants;
    }
    
    private void initializeScreenStateReceiver() {
        screenStateReceiver = new ScreenStateReceiver(new ScreenStateReceiver.ScreenStateListener() {
            @Override
            public void onScreenStateChanged(boolean isScreenOn) {
                Log.d(TAG, "Screen state changed: " + (isScreenOn ? "ON" : "OFF"));
                
                // Save screen state
                prefs.edit().putBoolean(KEY_SCREEN_STATE, isScreenOn).apply();
                
                // Update timer service
                updateTimerServiceState();
                
                // Emit event to React Native
                WritableMap params = Arguments.createMap();
                params.putBoolean("isScreenOn", isScreenOn);
                params.putLong("timestamp", System.currentTimeMillis());
                sendEvent(EVENT_SCREEN_STATE_CHANGED, params);
            }
        });
    }
    
    @ReactMethod
    public void startListening() {
        if (!isListening) {
            Log.d(TAG, "Starting to listen for screen state changes");
            
            IntentFilter filter = new IntentFilter();
            filter.addAction(Intent.ACTION_SCREEN_ON);
            filter.addAction(Intent.ACTION_SCREEN_OFF);
            
            try {
                reactContext.registerReceiver(screenStateReceiver, filter);
                isListening = true;
                Log.d(TAG, "Screen state receiver registered successfully");
            } catch (Exception e) {
                Log.e(TAG, "Failed to register screen state receiver", e);
            }
        }
    }
    
    @ReactMethod
    public void stopListening() {
        if (isListening) {
            try {
                reactContext.unregisterReceiver(screenStateReceiver);
                isListening = false;
                Log.d(TAG, "Screen state receiver unregistered");
            } catch (Exception e) {
                Log.e(TAG, "Failed to unregister screen state receiver", e);
            }
        }
    }
    
    @ReactMethod
    public void startTracking(Promise promise) {
        try {
            Log.d(TAG, "Starting timer tracking");
            
            // Save tracking state
            prefs.edit()
                .putBoolean(KEY_IS_TRACKING, true)
                .putLong(KEY_LAST_UPDATE, System.currentTimeMillis())
                .apply();
            
            // Start foreground service
            startTimerService();
            
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Failed to start tracking", e);
            promise.reject("START_TRACKING_ERROR", e.getMessage());
        }
    }
    
    @ReactMethod
    public void stopTracking(Promise promise) {
        try {
            Log.d(TAG, "Stopping timer tracking");
            
            // Save tracking state
            prefs.edit().putBoolean(KEY_IS_TRACKING, false).apply();
            
            // Stop foreground service
            stopTimerService();
            
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Failed to stop tracking", e);
            promise.reject("STOP_TRACKING_ERROR", e.getMessage());
        }
    }
    
    @ReactMethod
    public void addTimeCredits(int seconds, Promise promise) {
        try {
            Log.d(TAG, "Adding time credits: " + seconds + " seconds");
            
            long currentTime = getRemainingTimeInternal();
            long newTime = Math.max(0, currentTime + seconds);
            
            // Save new time
            prefs.edit()
                .putLong(KEY_REMAINING_TIME, newTime)
                .putLong(KEY_LAST_UPDATE, System.currentTimeMillis())
                .apply();
            
            // Update service if running
            if (isServiceRunning) {
                updateTimerService();
            }
            
            // Send update event
            sendTimerUpdateEvent();
            
            promise.resolve(newTime);
        } catch (Exception e) {
            Log.e(TAG, "Failed to add time credits", e);
            promise.reject("ADD_TIME_ERROR", e.getMessage());
        }
    }
    
    @ReactMethod
    public void getRemainingTime(Promise promise) {
        try {
            long remainingTime = getRemainingTimeInternal();
            promise.resolve((double) remainingTime);
        } catch (Exception e) {
            Log.e(TAG, "Failed to get remaining time", e);
            promise.reject("GET_TIME_ERROR", e.getMessage());
        }
    }
    
    @ReactMethod
    public void getTimerStatus(Promise promise) {
        try {
            WritableMap status = Arguments.createMap();
            status.putDouble("remainingTime", getRemainingTimeInternal());
            status.putBoolean("isTracking", prefs.getBoolean(KEY_IS_TRACKING, false));
            status.putDouble("debtTime", prefs.getLong(KEY_DEBT_TIME, 0));
            status.putBoolean("isScreenOn", prefs.getBoolean(KEY_SCREEN_STATE, true));
            status.putBoolean("isAppForeground", isAppInForeground());
            status.putLong("lastUpdate", prefs.getLong(KEY_LAST_UPDATE, 0));
            
            promise.resolve(status);
        } catch (Exception e) {
            Log.e(TAG, "Failed to get timer status", e);
            promise.reject("GET_STATUS_ERROR", e.getMessage());
        }
    }
    
    @ReactMethod
    public void notifyAppState(String state) {
        Log.d(TAG, "App state changed to: " + state);
        
        boolean isForeground = "app_foreground".equals(state);
        prefs.edit().putBoolean(KEY_APP_STATE, isForeground).apply();
        
        // Update timer service state
        updateTimerServiceState();
        
        // Emit event to React Native
        WritableMap params = Arguments.createMap();
        params.putString("state", state);
        params.putBoolean("isForeground", isForeground);
        params.putLong("timestamp", System.currentTimeMillis());
        sendEvent(EVENT_APP_STATE_CHANGED, params);
    }
    
    @ReactMethod
    public void resetTimer(Promise promise) {
        try {
            Log.d(TAG, "Resetting timer");
            
            prefs.edit()
                .putLong(KEY_REMAINING_TIME, 0)
                .putLong(KEY_DEBT_TIME, 0)
                .putBoolean(KEY_IS_TRACKING, false)
                .putLong(KEY_LAST_UPDATE, System.currentTimeMillis())
                .apply();
            
            stopTimerService();
            sendTimerUpdateEvent();
            
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Failed to reset timer", e);
            promise.reject("RESET_TIMER_ERROR", e.getMessage());
        }
    }
    
    private long getRemainingTimeInternal() {
        long savedTime = prefs.getLong(KEY_REMAINING_TIME, 0);
        long lastUpdate = prefs.getLong(KEY_LAST_UPDATE, System.currentTimeMillis());
        boolean isTracking = prefs.getBoolean(KEY_IS_TRACKING, false);
        
        if (!isTracking) {
            return savedTime;
        }
        
        // Calculate elapsed time since last update
        long currentTime = System.currentTimeMillis();
        long elapsedSeconds = (currentTime - lastUpdate) / 1000;
        
        // Only subtract time if conditions are met (background + screen on)
        if (shouldTimerRun()) {
            return Math.max(savedTime - elapsedSeconds, Long.MIN_VALUE / 2); // Allow negative time
        }
        
        return savedTime;
    }
    
    private boolean shouldTimerRun() {
        boolean isScreenOn = prefs.getBoolean(KEY_SCREEN_STATE, true);
        boolean isAppForeground = isAppInForeground();
        
        // Timer runs when app is in background AND screen is on
        return !isAppForeground && isScreenOn;
    }
    
    private boolean isAppInForeground() {
        // Check if app is in foreground using stored state
        return prefs.getBoolean(KEY_APP_STATE, true);
    }
    
    private void startTimerService() {
        if (!isServiceRunning) {
            Intent serviceIntent = new Intent(reactContext, TimerService.class);
            serviceIntent.setAction(TimerService.ACTION_START_TIMER);
            
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    reactContext.startForegroundService(serviceIntent);
                } else {
                    reactContext.startService(serviceIntent);
                }
                isServiceRunning = true;
                Log.d(TAG, "Timer service started");
            } catch (Exception e) {
                Log.e(TAG, "Failed to start timer service", e);
            }
        }
    }
    
    private void stopTimerService() {
        if (isServiceRunning) {
            Intent serviceIntent = new Intent(reactContext, TimerService.class);
            reactContext.stopService(serviceIntent);
            isServiceRunning = false;
            Log.d(TAG, "Timer service stopped");
        }
    }
    
    private void updateTimerService() {
        if (isServiceRunning) {
            Intent serviceIntent = new Intent(reactContext, TimerService.class);
            serviceIntent.setAction(TimerService.ACTION_UPDATE_TIMER);
            
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    reactContext.startForegroundService(serviceIntent);
                } else {
                    reactContext.startService(serviceIntent);
                }
            } catch (Exception e) {
                Log.e(TAG, "Failed to update timer service", e);
            }
        }
    }
    
    private void updateTimerServiceState() {
        if (isServiceRunning) {
            Intent serviceIntent = new Intent(reactContext, TimerService.class);
            serviceIntent.setAction(TimerService.ACTION_STATE_CHANGED);
            serviceIntent.putExtra("shouldRun", shouldTimerRun());
            
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    reactContext.startForegroundService(serviceIntent);
                } else {
                    reactContext.startService(serviceIntent);
                }
            } catch (Exception e) {
                Log.e(TAG, "Failed to update timer service state", e);
            }
        }
    }
    
    private void sendTimerUpdateEvent() {
        WritableMap params = Arguments.createMap();
        params.putDouble("remainingTime", getRemainingTimeInternal());
        params.putBoolean("isTracking", prefs.getBoolean(KEY_IS_TRACKING, false));
        params.putDouble("debtTime", prefs.getLong(KEY_DEBT_TIME, 0));
        params.putBoolean("isScreenOn", prefs.getBoolean(KEY_SCREEN_STATE, true));
        params.putBoolean("isAppForeground", isAppInForeground());
        params.putLong("timestamp", System.currentTimeMillis());
        
        sendEvent(EVENT_TIMER_UPDATE, params);
    }
    
    private void sendEvent(String eventName, @Nullable WritableMap params) {
        if (reactContext.hasActiveCatalystInstance()) {
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
        }
    }
    
    // TimerServiceCallback implementation
    @Override
    public void onTimerUpdate(long remainingTime, long debtTime) {
        // Update stored values
        prefs.edit()
            .putLong(KEY_REMAINING_TIME, remainingTime)
            .putLong(KEY_DEBT_TIME, debtTime)
            .putLong(KEY_LAST_UPDATE, System.currentTimeMillis())
            .apply();
        
        // Send update to React Native
        sendTimerUpdateEvent();
    }
    
    @Override
    public void onInvalidate() {
        super.onInvalidate();
        stopListening();
        stopTimerService();
    }
}

File: android/app/src/main/java/com/brainbites/timer/BrainBitesTimerModule.kt
// android/app/src/main/java/com/brainbites/timer/BrainBitesTimerModule.kt
package com.brainbites.timer

import android.app.KeyguardManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.os.PowerManager
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

class BrainBitesTimerModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "BrainBitesTimerModule"
        private const val MODULE_NAME = "BrainBitesTimer"
        
        // Event names
        const val EVENT_TIMER_UPDATE = "BrainBitesTimerUpdate"
        const val EVENT_SCREEN_STATE_CHANGED = "BrainBitesScreenStateChanged"
        const val EVENT_APP_STATE_CHANGED = "BrainBitesAppStateChanged"
        
        // SharedPreferences keys
        private const val PREFS_NAME = "BrainBitesTimerPrefs"
        private const val KEY_REMAINING_TIME = "remaining_time"
        private const val KEY_DEBT_TIME = "debt_time"
        private const val KEY_LAST_UPDATE = "last_update"
        private const val KEY_IS_TRACKING = "is_tracking"
    }

    private var timerStorage: TimerStorage? = null
    private var notificationManager: TimerNotificationManager? = null
    private var powerManager: PowerManager? = null
    private var keyguardManager: KeyguardManager? = null
    private var sharedPrefs: SharedPreferences? = null
    
    // Current timer state
    private var remainingTimeSeconds = 0
    private var debtTimeSeconds = 0
    private var isTracking = false
    private var isAppForeground = true
    private var isScreenOn = true
    private var lastUpdateTime = System.currentTimeMillis()
    
    // Screen state receiver
    private var screenStateReceiver: ScreenStateReceiver? = null
    private var isListening = false

    override fun getName(): String = MODULE_NAME

    override fun initialize() {
        super.initialize()
        try {
            Log.d(TAG, "Initializing BrainBitesTimerModule")
            
            // Initialize system services
            powerManager = reactContext.getSystemService(Context.POWER_SERVICE) as PowerManager
            keyguardManager = reactContext.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
            sharedPrefs = reactContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            
            // Initialize our custom classes
            timerStorage = TimerStorage(reactContext)
            notificationManager = TimerNotificationManager(reactContext)
            
            // Load saved state
            loadTimerState()
            
            Log.d(TAG, "BrainBitesTimerModule initialized successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize BrainBitesTimerModule", e)
        }
    }

    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        stopListening()
    }

    override fun getConstants(): MutableMap<String, Any> {
        return hashMapOf(
            "EVENT_TIMER_UPDATE" to EVENT_TIMER_UPDATE,
            "EVENT_SCREEN_STATE_CHANGED" to EVENT_SCREEN_STATE_CHANGED,
            "EVENT_APP_STATE_CHANGED" to EVENT_APP_STATE_CHANGED
        )
    }

    @ReactMethod
    fun startListening() {
        if (isListening) return
        
        try {
            Log.d(TAG, "Starting to listen for screen state changes")
            
            screenStateReceiver = ScreenStateReceiver { isScreenOn ->
                this.isScreenOn = isScreenOn
                updateTimerState()
                sendScreenStateEvent(isScreenOn)
            }
            
            val filter = IntentFilter().apply {
                addAction(Intent.ACTION_SCREEN_ON)
                addAction(Intent.ACTION_SCREEN_OFF)
                addAction(Intent.ACTION_USER_PRESENT)
            }
            
            reactContext.registerReceiver(screenStateReceiver, filter)
            isListening = true
            
            // Get initial screen state
            isScreenOn = powerManager?.isInteractive ?: true
            
            Log.d(TAG, "Started listening, initial screen state: $isScreenOn")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start listening", e)
        }
    }

    @ReactMethod
    fun stopListening() {
        if (!isListening) return
        
        try {
            Log.d(TAG, "Stopping screen state listener")
            screenStateReceiver?.let { receiver ->
                reactContext.unregisterReceiver(receiver)
            }
            screenStateReceiver = null
            isListening = false
        } catch (e: Exception) {
            Log.e(TAG, "Failed to stop listening", e)
        }
    }

    @ReactMethod
    fun startTracking(promise: Promise) {
        try {
            Log.d(TAG, "Starting timer tracking")
            isTracking = true
            lastUpdateTime = System.currentTimeMillis()
            saveTimerState()
            
            // Start the timer service
            val serviceIntent = Intent(reactContext, TimerService::class.java).apply {
                action = TimerService.ACTION_START_TIMER
            }
            reactContext.startForegroundService(serviceIntent)
            
            updateTimerState()
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start tracking", e)
            promise.reject("START_TRACKING_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun stopTracking(promise: Promise) {
        try {
            Log.d(TAG, "Stopping timer tracking")
            isTracking = false
            saveTimerState()
            
            // Stop the timer service
            val serviceIntent = Intent(reactContext, TimerService::class.java).apply {
                action = TimerService.ACTION_STOP_TIMER
            }
            reactContext.startService(serviceIntent)
            
            updateTimerState()
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to stop tracking", e)
            promise.reject("STOP_TRACKING_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun addTimeCredits(seconds: Int, promise: Promise) {
        try {
            Log.d(TAG, "Adding $seconds seconds to timer")
            
            if (debtTimeSeconds > 0) {
                // First pay off debt
                val debtPayment = minOf(seconds, debtTimeSeconds)
                debtTimeSeconds -= debtPayment
                val remainingCredits = seconds - debtPayment
                
                if (remainingCredits > 0) {
                    remainingTimeSeconds += remainingCredits
                }
                
                Log.d(TAG, "Paid off $debtPayment debt, added $remainingCredits to remaining time")
            } else {
                // Add directly to remaining time
                remainingTimeSeconds += seconds
            }
            
            saveTimerState()
            updateTimerState()
            
            promise.resolve(remainingTimeSeconds)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to add time credits", e)
            promise.reject("ADD_TIME_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun getRemainingTime(promise: Promise) {
        try {
            promise.resolve(remainingTimeSeconds)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get remaining time", e)
            promise.reject("GET_TIME_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun getTimerStatus(promise: Promise) {
        try {
            val statusMap = WritableNativeMap().apply {
                putInt("remainingTime", remainingTimeSeconds)
                putBoolean("isTracking", isTracking)
                putInt("debtTime", debtTimeSeconds)
                putBoolean("isScreenOn", isScreenOn)
                putBoolean("isAppForeground", isAppForeground)
                putDouble("lastUpdate", lastUpdateTime.toDouble())
            }
            promise.resolve(statusMap)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get timer status", e)
            promise.reject("GET_STATUS_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun notifyAppState(state: String) {
        try {
            Log.d(TAG, "App state changed to: $state")
            isAppForeground = when (state.lowercase()) {
                "foreground", "active" -> true
                "background", "inactive" -> false
                else -> isAppForeground
            }
            
            updateTimerState()
            sendAppStateEvent(state, isAppForeground)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to handle app state change", e)
        }
    }

    @ReactMethod
    fun resetTimer(promise: Promise) {
        try {
            Log.d(TAG, "Resetting timer")
            remainingTimeSeconds = 0
            debtTimeSeconds = 0
            isTracking = false
            lastUpdateTime = System.currentTimeMillis()
            
            saveTimerState()
            
            // Stop timer service
            val serviceIntent = Intent(reactContext, TimerService::class.java).apply {
                action = TimerService.ACTION_STOP_TIMER
            }
            reactContext.startService(serviceIntent)
            
            updateTimerState()
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to reset timer", e)
            promise.reject("RESET_ERROR", e.message, e)
        }
    }

    // Internal methods
    private fun loadTimerState() {
        try {
            sharedPrefs?.let { prefs ->
                remainingTimeSeconds = prefs.getInt(KEY_REMAINING_TIME, 0)
                debtTimeSeconds = prefs.getInt(KEY_DEBT_TIME, 0)
                isTracking = prefs.getBoolean(KEY_IS_TRACKING, false)
                lastUpdateTime = prefs.getLong(KEY_LAST_UPDATE, System.currentTimeMillis())
                
                Log.d(TAG, "Loaded timer state: remaining=$remainingTimeSeconds, debt=$debtTimeSeconds, tracking=$isTracking")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load timer state", e)
        }
    }

    private fun saveTimerState() {
        try {
            sharedPrefs?.edit()?.apply {
                putInt(KEY_REMAINING_TIME, remainingTimeSeconds)
                putInt(KEY_DEBT_TIME, debtTimeSeconds)
                putBoolean(KEY_IS_TRACKING, isTracking)
                putLong(KEY_LAST_UPDATE, System.currentTimeMillis())
                apply()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save timer state", e)
        }
    }

    private fun updateTimerState() {
        try {
            // Update notification
            notificationManager?.updateNotification(
                remainingTimeSeconds = remainingTimeSeconds,
                debtTimeSeconds = debtTimeSeconds,
                isTracking = shouldTrackTime()
            )
            
            // Send event to React Native
            sendTimerUpdateEvent()
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to update timer state", e)
        }
    }

    private fun shouldTrackTime(): Boolean {
        return isTracking && !isAppForeground && isScreenOn && !(keyguardManager?.isKeyguardLocked ?: false)
    }

    private fun sendTimerUpdateEvent() {
        try {
            val eventData = WritableNativeMap().apply {
                putInt("remainingTime", remainingTimeSeconds)
                putBoolean("isTracking", shouldTrackTime())
                putInt("debtTime", debtTimeSeconds)
                putBoolean("isScreenOn", isScreenOn)
                putBoolean("isAppForeground", isAppForeground)
                putDouble("timestamp", System.currentTimeMillis().toDouble())
            }
            
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(EVENT_TIMER_UPDATE, eventData)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send timer update event", e)
        }
    }

    private fun sendScreenStateEvent(isScreenOn: Boolean) {
        try {
            val eventData = WritableNativeMap().apply {
                putBoolean("isScreenOn", isScreenOn)
                putDouble("timestamp", System.currentTimeMillis().toDouble())
            }
            
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(EVENT_SCREEN_STATE_CHANGED, eventData)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send screen state event", e)
        }
    }

    private fun sendAppStateEvent(state: String, isForeground: Boolean) {
        try {
            val eventData = WritableNativeMap().apply {
                putString("state", state)
                putBoolean("isForeground", isForeground)
                putDouble("timestamp", System.currentTimeMillis().toDouble())
            }
            
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(EVENT_APP_STATE_CHANGED, eventData)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send app state event", e)
        }
    }
}